platform/android/src/jni.cpp

```diff
@@ -1620,6 +1753,363 @@ void JNICALL nativeRemoveCustomLayer(JNIEnv *env, jobject obj, jlong nativeMapVi
 }
 
+// Offline calls begin
+
+jlong JNICALL createDefaultFileSource(JNIEnv *env, jobject obj, jstring cachePath_, jstring assetRoot_, jlong maximumCacheSize) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "createDefaultFileSource");
+    std::string cachePath = std_string_from_jstring(env, cachePath_);
+    std::string assetRoot = std_string_from_jstring(env, assetRoot_);
+    mbgl::DefaultFileSource *defaultFileSource = new mbgl::DefaultFileSource(cachePath, assetRoot, maximumCacheSize);
+    jlong defaultFileSourcePtr = reinterpret_cast<jlong>(defaultFileSource);
+    return defaultFileSourcePtr;
+}
+
+void JNICALL setAccessToken(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr, jstring accessToken_) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "setAccessToken");
+    assert(defaultFileSourcePtr != 0);
+    std::string accessToken = std_string_from_jstring(env, accessToken_);
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    defaultFileSource->setAccessToken(accessToken);
+}
+
+jstring JNICALL getAccessToken(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "getAccessToken");
+    assert(defaultFileSourcePtr != 0);
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    std::string accessToken = defaultFileSource->getAccessToken();
+    return std_string_to_jstring(env, accessToken);
+}
+
+void JNICALL listOfflineRegions(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr, jobject listCallback) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "listOfflineRegions");
+
+    // Checks
+    assert(defaultFileSourcePtr != 0);
+    if (listCallback == nullptr) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Callback has to be set.");
+        return;
+    }
+
+    // Makes sure the callback doesn't get GC'ed
+    listCallback = reinterpret_cast<jobject>(env->NewGlobalRef(listCallback));
+
+    // Launch listCallback
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    defaultFileSource->listOfflineRegions([defaultFileSourcePtr, listCallback](std::exception_ptr error, mbgl::optional<std::vector<mbgl::OfflineRegion>> regions) {
+
+        // Reattach, the callback comes from a different thread
+        JNIEnv *env2;
+        jboolean renderDetach = attach_jni_thread(theJVM, &env2, "Offline Thread");
+        if (renderDetach) {
+            mbgl::Log::Debug(mbgl::Event::JNI, "Attached.");
+        }
+
+        if (error) {
+            std::string message = mbgl::util::toString(error);
+            env2->CallVoidMethod(listCallback, listOnErrorMethodId, std_string_to_jstring(env2, message));
+        } else if (regions) {
+            // Build the OfflineManager object
+            jobject jmanager = env2->NewObject(offlineManagerClass, offlineManagerClassConstructorId);
```

Comment:
Why is this allocating a new manager? Shouldn't it be reusing the `OfflineManager` on which `listOfflineRegions()` was called?